lua/nvim-tree/explorer/filters.lua

```diff
@@ -164,9 +164,8 @@ function M.setup(opts)
   M.exclude_list = opts.filters.exclude
 
   local custom_filter = opts.filters.custom
-
   if type(custom_filter) == "function" then
-    M.config.filter_custom = custom_filter
+    M.ignore_list[custom_filter] = true
```

Comment:
Should we be using ignore_list at all; can't we just use `M.config.custom_filter`  directly?

Unfortunately I'm out of time until the weekend. I'll give it a solid test then.

Reply:
By `M.config.custom_filter`, I assume you mean `M.config.filter_custom`?

It looks like `M.config.filter_custom` is a boolean, for whether all custom filters are enabled.
Adding the single custom filter function to `M.ignore_list` would be consistent with adding each item of the custom filter table to it, wouldn't it? As `M.ignore_list` is a set, with each item checked in the loop in the `custom()`.

Perhaps I should set `M.ignore_list` to `opts.filters.custom` (a function), and check the type in `custom()`, if this is more intuitive or performant.

Reply:
> It looks like M.config.filter_custom is a boolean, for whether all custom filters are enabled.

Right you are, please accept my apologies.

> Perhaps I should set `M.ignore_list` to `opts.filters.custom` (a function), and check the type in `custom()`, if this is more intuitive or performant.

That is more intuitive, however it's not good practice to use a member with variable types.

Can we please go one further and explicitly separate, something like

```diff
diff --git a/lua/nvim-tree/explorer/filters.lua b/lua/nvim-tree/explorer/filters.lua
index f3beb61..24747c2 100644
--- a/lua/nvim-tree/explorer/filters.lua
+++ b/lua/nvim-tree/explorer/filters.lua
@@ -4,6 +4,7 @@ local marks = require "nvim-tree.marks"
 local M = {
   ignore_list = {},
   exclude_list = {},
+  custom_function = nil,
 }

 ---@param path string
@@ -84,18 +85,17 @@ local function custom(path)

   local basename = utils.path_basename(path)

+  -- filter by user's custom function
+  if M.custom_function and M.custom_function(path) then
+    return true
+  end
+
   -- filter custom regexes
   local relpath = utils.path_relative(path, vim.loop.cwd())
   for pat, _ in pairs(M.ignore_list) do
-    if type(pat) == "function" then
-      if pat(path) then
-        return true
-      end
-    else
       if vim.fn.match(relpath, pat) ~= -1 or vim.fn.match(basename, pat) ~= -1 then
         return true
       end
-    end
   end

   local idx = path:match ".+()%.[^.]+$"
@@ -160,7 +160,7 @@ function M.setup(opts)

   local custom_filter = opts.filters.custom
   if type(custom_filter) == "function" then
-    M.ignore_list[custom_filter] = true
+    M.custom_function = custom_filter
   else
     if custom_filter and #custom_filter > 0 then
       for _, filter_name in pairs(custom_filter) do
```

Reply:
That looks like a better solution.
Though this won't let a user define a both a function and a pattern, but I don't think this is a problem, as a custom function can of course contain a pattern check.

```lua
custom = function(path)
    if vim.fn.match(path, ".*\\.lock") ~= -1 then -- Could be a specific function like substring
        return true
    end
    return is_binary(path)
end
```

Previously:

```lua
custom = {
    "*.lock",
    function(path) -- Or simply pass `is_binary`
        return is_binary(path)
    end,
}
```

Reply:
Is it possible to go back to expecting a `table<function | string>`, instead of a `function | table<string>`?
A singular function would just be wrapped in a table: `custom = { is_binary }`.

If this is not preferable, I do not mind.

Reply:
> Though this won't let a user define a both a function and a pattern, but I don't think this is a problem, as a custom function can of course contain a pattern check.

> Is it possible to go back to expecting a table<function | string>, instead of a function | table<string>?
A singular function would just be wrapped in a table: custom = { is_binary }.

Sorry, I'd prefer we keep it in the style of the other similar code - one (fast) function only.